SCClassLibrary/Common/GUI/PlusGUI/Control/ServerPlusGUI.sc

```diff
@@ -1,33 +1,42 @@
 ServerTreeView {
 	var <server, <bounds, viewParent, <window, resp, updateFunc, updater, tabSize = 25, treeViewStatus;
 
-	*new { |server, bounds=(Rect(128, 64, 400, 400)), parent|
-		if(bounds.asRect.width < 400) {
-			var bnds = bounds.asRect;
-			"The width value you set will be changed to 400, the minimum width.".postln;
-			bounds = Rect(bnds.left, bnds.top, 400, bnds.height);
-		};
+	*new { |server, bounds, parent|
+		server = server ? Server.default;
 		^super.newCopyArgs(server).makeWindow(bounds, parent)
 	}
 
 	makeWindow { |bounds, parent|
-		if(parent == nil) {
-			window = Window.new(this.asString, bounds, scroll:true);
-			viewParent = window;
+		bounds = if(bounds.isNil) {
+			Rect(128, 64, 400, 400)
 		} {
-			if (parent.name.contains(" Node Tree")) {
-				window = parent
+			if(bounds.asRect.width < 400) {
+				var bnds = bounds.asRect;
+				"The width value you set will be changed to 400, the minimum width.".postln;
+				Rect(bnds.left, bnds.top, 400, bnds.height);
 			} {
+				bounds
+			}
+		};
+		if(parent.isNil) {
+			window = Window(this.asString, bounds, scroll:true);
+		} {
+			if(parent.view.asString == "a TopView") {
```

Comment:
can you explain what this does and why this is necessary?

Reply:
The part 
```
if(parent.view.asString == "a TopView") {
				var name = parent.name;
				(
					"The Window should be created with scroll argumrnt value true. e.g.:" ++
					"Window.(" ++ "title".quote ++ ", scroll: true)"
				).postln;
				window = Window(parent.name, bounds, scroll:true);
				parent.close;
			} {
				window = parent;
			}
```
is used to prevent the following errors:

```
-> a Window
ERROR: Message 'hasHorizontalScroller_' not understood.
RECEIVER:
Instance of TopView {    (0x12b7f3918, gc=34, fmt=00, flg=00, set=06)
  instance variables [35]
    qObject : RawPointer 0x6000051913e0
    finalizer : instance of Finalizer (0x12cbbd6a8, size=2, set=1)
    virtualSlots : instance of Array (0x12b94cc78, size=1, set=2)
    wasRemoved : false
    font : nil
    resize : Integer 1
    alpha : Float 1.000000   00000000 3FF00000
    decorator : nil
    layout : nil
    userCanClose : true
    deleteOnClose : true
    action : nil
    mouseDownAction : nil
    mouseUpAction : nil
    mouseEnterAction : nil
    mouseLeaveAction : nil
    mouseMoveAction : nil
    mouseOverAction : nil
    mouseWheelAction : nil
    keyDownAction : nil
    keyUpAction : nil
    keyModifiersChangedAction : nil
    keyTyped : nil
    focusGainedAction : nil
    focusLostAction : nil
    dragLabel : nil
    beginDragAction : nil
    canReceiveDragHandler : nil
    receiveDragHandler : nil
    toFrontAction : nil
    endFrontAction : nil
    onClose : nil
    onResize : nil
    onMove : nil
    window : instance of Window (0x122b0c588, size=5, set=3)
}
ARGS:
   false
PATH: /Users/prko/Dropbox/prko/__myDocs/Writings/Making Sound using Open Sources/mixed/dev - Bleeding edge/SuperCollider.app/Contents/Resources/HelpSource/Tutorials/aKoreanGuide/test.scd
CALL STACK:
	DoesNotUnderstandError:reportError
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand
		arg this = <instance of TopView>
		arg selector = 'hasHorizontalScroller_'
		arg args = [*1]
	ServerTreeView:makeWindow
		arg this = <instance of ServerTreeView>
		arg bounds = <instance of Rect>
		arg parent = <instance of Window>
	Server:plotTreeView
		arg this = <instance of Server>
		arg interval = 0.5
		arg parent = <instance of Window>
		arg actionIfFail = nil
	< closed FunctionDef >  (no arguments or variables)
	Interpreter:interpretPrintCmdLine
		arg this = <instance of Interpreter>
		var res = nil
		var func = <instance of Function>
		var code = "~v = s.plotTreeView(parent: ~u)"
		var doc = nil
		var ideClass = <instance of Meta_ScIDE>
	Process:interpretPrintCmdLine
		arg this = <instance of Main>
^^ ERROR: Message 'hasHorizontalScroller_' not understood.
RECEIVER: a TopView
```
However, this solution has a significant issue: the connection between the variable and its window is disconnected.
```
~u = Window();
~v = s.plotTreeView(parent: ~u)
~w = s.plotTreeView(parent: ~u)
~u.alwaysOnTop_(true)
```
Would it be better to change this part as follows (post an informative text to solve the error).?
```
if(parent.isNil) {
			window = Window(this.asString, bounds, scroll:true);
		} {
			if(parent.view.asString == "a TopView") {
				(
					"The Window should be created with scroll argumrnt value true. e.g.:" ++
					"Window.(" ++ "title".quote ++ ", scroll: true)"
				).postln;
			};
			window = parent
		};
```

Reply:
Now I understand.

This is not the right way to do it (in general), because it makes the behaviour depend on the title of a window which can be anything.

To solve it, you just look what classes implement `hasHorizontalScroller_`. It is only `AbstractScroll`. 

Using `AbstractScroll.dumpClassSubtree` you can see all classes that implement it.

I think in this case it probably is `ScrollView`.

A better way is below, where the method is called, to do a check:

```supercollider
if(window.view.respondsTo(\hasHorizontalScroller_)) { window.view.hasHorizontalScroller_(false) };
```

But actually, I think this specification should not happen here at all, but where the new window is created: in line 22.

Reply:
… then you can drop the whole extra branch.

Reply:
Thanks!